Mass monkeypatching leads to boto3 incompatibility and infinite recursion in pytest (when used as lib)

<!-- 
If you have a general question about how to use Locust, please check Stack Overflow first https://stackoverflow.com/questions/tagged/locust

You can also ask new questions on SO, https://stackoverflow.com/questions/ask just remember to tag your question with "locust". Do not immediately post your issue here after posting to SO, wait for an answer there instead.

Use this form only for reporting actual bugs in Locust. Remember, the developers of Locust are unpaid volunteers, so make sure you have tried everything you can think of before filing a bug :) 

Always make sure you are running an up to date Locust version (pip3 install -U locust)
-->

### Describe the bug
<!-- A clear and concise description of what the bug is -->
I know this [issue](https://github.com/gevent/gevent/issues/1016) has been referenced before, but it seems that re-patching all modules with the exception of Python's native SSL with:
```
from gevent import monkey
monkey.patch_all(ssl=False)
```
in the beginning of a Python test file doesn't lead to any positive results when using locust as library in pytest if boto3 is also installed.

I wanted to introduce pytest load testing in a PR to a repo that I cloned. Chose locust for integration and extension capabilities.
Wrote a simple [function](https://github.com/inpv/cvat/blob/inpv_test_task_mass_user_register/tests/python/rest_api/test_load_register.py) to use locust as library based on the example in the docs, checked that the test works as expected, but if I try to launch all REST API tests with locust installed, half of them fail because of mass monkeypatching that locust introduces in the [init](https://github.com/locustio/locust/blob/b64b9a6082451be74f94bf96406315fa819fcd1c/locust/__init__.py#L11) file. So I disabled ssl monkeypatching explicitly like I wrote above. Now this
```
.env/lib/python3.9/site-packages/locust/__init__.py:11
  /home/inpv/Development/cvat/.env/lib/python3.9/site-packages/locust/__init__.py:11: MonkeyPatchWarning: Monkey-patching ssl after ssl has already been imported may lead to errors, including RecursionError on Python 3.6. It may also silently lead to incorrect behaviour on Python 3.7. Please monkey-patch earlier. See https://github.com/gevent/gevent/issues/1016. Modules that had direct imports (NOT patched): ['urllib3.util (/home/inpv/Development/cvat/.env/lib/python3.9/site-packages/urllib3/util/__init__.py)', 'urllib3.util.ssl_ (/home/inpv/Development/cvat/.env/lib/python3.9/site-packages/urllib3/util/ssl_.py)']. 
    monkey.patch_all()
```
warning is gone, but there is another:
```
.env/lib/python3.9/site-packages/locust/__init__.py:11
  /home/inpv/Development/cvat/.env/lib/python3.9/site-packages/locust/__init__.py:11: MonkeyPatchWarning: Patching more than once will result in the union of all True parameters being patched
    monkey.patch_all()
```

and the recursion error still is there. If I try to launch a REST API test that otherwise passes without a `locust` import, regardless of SSL is re-patched or not, the [test](https://github.com/inpv/cvat/blob/inpv_test_task_mass_user_register/tests/python/rest_api/test_resource_import_export.py) that utilizes AWS s3 buckets on `boto3` will yield an error like this:
```
tests/python/rest_api/test_resource_import_export.py:290: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
tests/python/rest_api/test_resource_import_export.py:75: in wrapper
    client = define_client()
tests/python/rest_api/test_resource_import_export.py:45: in define_client
    s3 = boto3.resource(
.env/lib/python3.9/site-packages/boto3/__init__.py:102: in resource
    return _get_default_session().resource(*args, **kwargs)
.env/lib/python3.9/site-packages/boto3/session.py:384: in resource
    client = self.client(
.env/lib/python3.9/site-packages/boto3/session.py:258: in client
    return self._session.create_client(
.env/lib/python3.9/site-packages/botocore/session.py:847: in create_client
    client = client_creator.create_client(
.env/lib/python3.9/site-packages/botocore/client.py:86: in create_client
    client_args = self._get_client_args(
.env/lib/python3.9/site-packages/botocore/client.py:355: in _get_client_args
    return args_creator.get_client_args(
.env/lib/python3.9/site-packages/botocore/args.py:99: in get_client_args
    endpoint = endpoint_creator.create_endpoint(
.env/lib/python3.9/site-packages/botocore/endpoint.py:293: in create_endpoint
    http_session = http_session_cls(
.env/lib/python3.9/site-packages/botocore/httpsession.py:203: in __init__
    self._manager = PoolManager(**self._get_pool_manager_kwargs())
.env/lib/python3.9/site-packages/botocore/httpsession.py:222: in _get_pool_manager_kwargs
    'ssl_context': self._get_ssl_context(),
.env/lib/python3.9/site-packages/botocore/httpsession.py:231: in _get_ssl_context
    return create_urllib3_context()
.env/lib/python3.9/site-packages/botocore/httpsession.py:77: in create_urllib3_context
    context.options |= options
/usr/lib/python3.9/ssl.py:603: in options
    super(SSLContext, SSLContext).options.__set__(self, value)
/usr/lib/python3.9/ssl.py:603: in options
    super(SSLContext, SSLContext).options.__set__(self, value)
E   RecursionError: maximum recursion depth exceeded while calling a Python object
!!! Recursion detected (same locals & position)
```
Because of this error the CI build fails with infinite recursion and I will never pass the `isort` linter checks with such a setup (for a good reason, actually). Is locust incompatible with boto3 using AWS s3 buckets or am I missing something? Why is SSL context creation called twice?

P.S. I've seen people face similar [errors](https://github.com/boto/boto3/issues/3365) on boto3 with different modules with no resolution but patching boto3 themselves, yet I decided to create the report here because imo mass monkeypatching may lead to this infinite recursion problem.

### Expected behavior
<!-- Tell us what you think should happen -->
Locust and gevent usage doesn't corrupt tests that utilize boto3 AWS s3 buckets.

### Actual behavior
<!-- Tell us what happens instead. Include screenshots if this an issue with the GUI. -->
The tests with boto3 usage fail with infinite recursion errors.


### Steps to reproduce
<!-- Please provide a minimal reproducible code example (https://stackoverflow.com/help/minimal-reproducible-example) --> 
Install latest locust
Install boto3 and botocore
Create a test with an AWS s3 bucket usage
Launch this test

### Environment

- OS: Arch Linux latest
- Python version: 3.9 in venv
- Locust version: (please dont file issues for anything but the most recent release or prerelease builds) 2.13.0
- Locust command line that you ran: usage as library, see links above
- Locust file contents (anonymized if necessary): usage as library, see links above

--- a/docs/use-as-lib.rst
+++ b/docs/use-as-lib.rst
@@ -49,6 +49,11 @@ to view the stats, and to control the runner (e.g. start and stop load tests):
     env.create_web_ui()
     env.web_ui.greenlet.join()
 
+Skipping monkey patching
+========================
+
+Some packages such as boto3 may have incompatibility when using Locust as a library, where monkey patching is already applied. In this case monkey patching may be disabled by setting ``LOCUST_SKIP_MONKEY_PATCH=1`` as env variable.
+
 Full example
 ============
 

--- a/locust/__init__.py
+++ b/locust/__init__.py
@@ -12,7 +12,8 @@ if os.getenv("LOCUST_PLAYWRIGHT", None):
 
 from gevent import monkey
 
-monkey.patch_all()
+if not os.getenv("LOCUST_SKIP_MONKEY_PATCH", None):
+    monkey.patch_all()
 
 from ._version import version as __version__
 from .contrib.fasthttp import FastHttpUser

